src/ObjCRuntime/Registrar.cs

```diff
@@ -159,6 +160,44 @@ internal class ObjCType {
 
 			public bool IsCategory { get { return CategoryAttribute != null; } }
 
+#if MTOUCH || MMP
+			HashSet<ObjCType> all_protocols;
+			// This contains all protocols in the type hierarchy.
+			// Given a type T that implements a protocol with super protocols:
+			// class T : NSObject, IProtocol2 {}
+			// [Protocol]
+			// interface IP1 {}
+			// [Protocol]
+			// interface IP2 : IP1 {}
+			// This property will contain both IP1 and IP2. The Protocols property only contains IP2.
+			public IEnumerable<ObjCType> AllProtocols {
+				get {
+					if (Protocols == null || Protocols.Length == 0)
+						return null;
```

Comment:
How important is the null return vs returning an empty`IEnumerable`?
It strikes me that you could also do:
```
bool protocolsRetrieved = false;
HashSet<ObjCType> all_protocols = new HashSet<ObjCType>();
public IEnumerable<ObjCType> AllProtocols {
    if (protocolsRetrieved)
        return all_protocols;
    if (Protocols != null && Protocols.Length > 0) {
        // code to fetch all the protocols here
    }
    protocolsRetrieved = true;
    return all_protocols;
}

Then you don't need the null tests on this type ever.

Reply:
@stephen-hawley because an empty `IEnumerable` uses more memory than null, and this is something that 99.9% of the time will be null (or empty).